pkg/syncer/upsync/upsync_cleanup_controller.go

```diff
@@ -0,0 +1,197 @@
+
+	"github.com/kcp-dev/kcp/pkg/indexers"
+	ddsif "github.com/kcp-dev/kcp/pkg/informer"
+	"github.com/kcp-dev/kcp/pkg/logging"
+	syncerindexers "github.com/kcp-dev/kcp/pkg/syncer/indexers"
+	"github.com/kcp-dev/kcp/pkg/syncer/shared"
+)
+
+const cleanupControllerName = "kcp-resource-upsyncer-cleanup"
+
+// NewUpSyncer returns a new controller which upsyncs, through the Upsyncer virtual workspace, downstream resources
+// which are part of the upsyncable resource types (fixed limited list for now), and provide
+// the following labels:
+//   - internal.workload.kcp.io/cluster: <sync target key>
+//   - state.workload.kcp.io/<sync target key>: Upsync
+//
+// and optionally, for cluster-wide resources, the `kcp.io/namespace-locator` annotation
+// filled with the information necessary identify the upstream workspace to upsync to.
+func NewUpSyncerCleanupController(syncerLogger logr.Logger, syncTargetClusterName logicalcluster.Name,
+	syncTargetName string, syncTargetUID types.UID, syncTargetKey string,
+	upstreamClusterClient kcpdynamic.ClusterInterface,
+	ddsifForUpstreamUpsyncer *ddsif.DiscoveringDynamicSharedInformerFactory,
+	ddsifForDownstream *ddsif.GenericDiscoveringDynamicSharedInformerFactory[cache.SharedIndexInformer, cache.GenericLister, informers.GenericInformer],
+) (*cleanupController, error) {
+	c := &cleanupController{
+		queue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), cleanupControllerName),
+		cleanupReconciler: cleanupReconciler{
+			getUpstreamClient: func(clusterName logicalcluster.Name) (dynamic.Interface, error) {
+				return upstreamClusterClient.Cluster(clusterName.Path()), nil
+			},
+			getDownstreamLister: func(gvr schema.GroupVersionResource) (cache.GenericLister, error) {
+				informers, notSynced := ddsifForDownstream.Informers()
+				informer, ok := informers[gvr]
+				if !ok {
+					if shared.ContainsGVR(notSynced, gvr) {
+						return nil, fmt.Errorf("informer for gvr %v not synced in the downstream informer factory", gvr)
+					}
+					return nil, fmt.Errorf("gvr %v should be known in the downstream informer factory", gvr)
+				}
+				return informer.Lister(), nil
+			},
+			listDownstreamNamespacesByLocator: func(jsonLocator string) ([]*unstructured.Unstructured, error) {
+				nsInformer, err := ddsifForDownstream.ForResource(namespaceGVR)
+				if err != nil {
+					return nil, err
+				}
+				return indexers.ByIndex[*unstructured.Unstructured](nsInformer.Informer().GetIndexer(), syncerindexers.ByNamespaceLocatorIndexName, jsonLocator)
+			},
+
+			syncTargetName:        syncTargetName,
+			syncTargetClusterName: syncTargetClusterName,
+			syncTargetUID:         syncTargetUID,
+		},
+	}
+	logger := logging.WithReconciler(syncerLogger, controllerName)
+
+	ddsifForUpstreamUpsyncer.AddEventHandler(ddsif.GVREventHandlerFuncs{
+		AddFunc: func(gvr schema.GroupVersionResource, obj interface{}) {
+			if gvr == namespaceGVR {
+				return
+			}
+			c.enqueue(gvr, obj, logger)
```

Comment:
enqueuNamespace?

Reply:
why ?
I precisely don't want to enqueue namespaces